### PR TITLE
Add concatMap, Monad instance for Folds.

### PR DESCRIPTION
**Pull request by pranaysashank**

### CODE REVIEW

File: src/Streamly/Internal/Data/Fold/Types.hs [hunk 1]
```diff
@@ -652,19 +652,67 @@ longest :: -- Monad m =>
...
+-- 45
+--
+-- /Internal/
 --
 {-# INLINE concatMap #-}
-concatMap :: -- Monad m =>
-    (b -> Fold m a c) -> Fold m a b -> Fold m a c
-concatMap _func _fld = undefined
+concatMap :: Monad m => (b -> Fold m a c) -> Fold m a b -> Fold m a c
+concatMap f (Fold stepa initiala extracta) = Fold stepc initialc extractc
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Fold/Types.hs:670]: `initiala` and family feel a little weird. Maybe use `initialA` etc.?

File: src/Streamly/Internal/Data/Fold/Types.hs [hunk 2]
```diff
@@ -652,19 +652,67 @@ longest :: -- Monad m =>
...
+    initialc = do
+        r <- initiala
+        case r of
+            Partial s -> return $ Partial (B s)
+            Done b -> generateInnerFold (f b)
+
+    stepc (B s) a = do
+        r <- stepa s a
+        case r of
+            Partial s' -> return $ Partial (B s')
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Fold/Types.hs:681]:
```suggestion
            Partial s1 -> return $ Partial (B s1)
```

File: src/Streamly/Internal/Data/Fold/Types.hs [hunk 3]
```diff
@@ -652,19 +652,67 @@ longest :: -- Monad m =>
...
+            Done b    -> generateInnerFold (f b)
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Fold/Types.hs:682]:
```suggestion
            Done b -> generateInnerFold (f b)
```

File: src/Streamly/Internal/Data/Fold/Types.hs [hunk 4]
```diff
@@ -652,19 +652,67 @@ longest :: -- Monad m =>
...
+            Partial s -> e s
+            Done c -> return c
+
+instance Monad m => Monad (Fold m a) where
+    {-# INLINE return #-}
+    return = pure
+
+    -- | __Notice:__ In the next major release the behaviour of @<*>@ for folds
+    -- will be updated from 'teeWith' to 'splitWith' to reflect it's similarity
+    -- with parsers.
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Fold/Types.hs:713]: Is this notice required here?

Comment (pranaysashank) [reply src/Streamly/Internal/Data/Fold/Types.hs:713]: Copy paste error from the Applicative instance

**Review by adithyaov — COMMENTED** [review]
There does not seem to be any functional error. You can merge it after addressing the comments and checking for regressions.

**Comment by pranaysashank** [comment]
What regressions should I be checking for?

File: src/Streamly/Internal/Data/Fold/Types.hs [hunk 5]
```diff
@@ -652,19 +652,67 @@ longest :: -- Monad m =>
...
+        r <- stepa s a
+        case r of
+            Partial s' -> return $ Partial (B s')
+            Done b    -> generateInnerFold (f b)
+
+    stepc (C stepInner s extractInner) a = do
+        r <- stepInner s a
+        case r of
+            Partial sc -> return $ Partial (C stepInner sc extractInner)
+            Done c -> return $ Done c
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Fold/Types.hs:688]: factor out `return`.

File: src/Streamly/Internal/Data/Fold/Types.hs [hunk 6]
```diff
@@ -652,19 +652,67 @@ longest :: -- Monad m =>
...
+    extractc (B s) = do
+        r <- extracta s
+        initExtract (f r)
+    extractc (C _ sInner extractInner) = extractInner sInner
+
+    generateInnerFold (Fold step i e) = do
+        r <- i
+        case r of
+            Partial s -> return $ Partial (C step s e)
+            Done c -> return $ Done c
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Fold/Types.hs:699]: factor out return.

File: src/Streamly/Internal/Data/Fold/Types.hs [hunk 7]
```diff
@@ -652,19 +652,67 @@ longest :: -- Monad m =>
...
-    (b -> Fold m a c) -> Fold m a b -> Fold m a c
-concatMap _func _fld = undefined
+concatMap :: Monad m => (b -> Fold m a c) -> Fold m a b -> Fold m a c
+concatMap f (Fold stepa initiala extracta) = Fold stepc initialc extractc
+  where
+    initialc = do
+        r <- initiala
+        case r of
+            Partial s -> return $ Partial (B s)
+            Done b -> generateInnerFold (f b)
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Fold/Types.hs:676]: This is more like `initInnerFold`.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]
Looks good. Can you add a benchmark? You can just copy and paste from the parser benchmarks.

File: src/Streamly/Internal/Data/Fold/Types.hs [hunk 8]
```diff
@@ -674,19 +674,64 @@ longest :: -- Monad m =>
...
+        case r of
+            Partial s -> e s
+            Done c -> return c
+
+instance Monad m => Monad (Fold m a) where
+    {-# INLINE return #-}
+    return = pure
+
+    {-# INLINE (>>=) #-}
+    (>>=) = flip concatMap
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Fold/Types.hs:734]: We cannot have different behavior for Applicative and Monad instances. Applicative is parallel tee and monad is serial chaining. Let's just keep concatMap and remove the Monad instance for now.

**Review by harendra-kumar — APPROVED** [review]

**Pull request merged**